pkg/controller/uniteddeployment/statefulset_control.go

```diff
@@ -0,0 +1,385 @@
+	subset.Status.RevisionReplicas = calculateStatus(pods, set)
+
+	return subset, nil
+}
+
+func getCurrentPartition(replicas int32, pods []*corev1.Pod, revision string) int32 {
+	ordinalMarks := make([]bool, replicas)
+	for _, pod := range pods {
+		ordinal := getOrdinal(pod)
+		if ordinal >= replicas {
+			// unexpected
+			continue
+		}
+		if getRevision(&pod.ObjectMeta) == revision {
+			ordinalMarks[ordinal] = true
+		}
+	}
+
+	partition := replicas
+	for i := replicas - 1; i >= 0; i-- {
+		if ordinalMarks[i] {
+			partition = i
+		} else {
+			break
+		}
+	}
+
+	return partition
+}
+
+func (m *StatefulSetControl) deleteStuckPods(set *appsv1.StatefulSet, revision string, partition int32) error {
+	pods, err := m.getStatefulSetPods(set)
+	if err != nil {
+		return err
+	}
+
+	for i := range pods {
+		pod := pods[i]
+		if isPodStuck(pod, revision, partition) {
+			klog.V(2).Infof("Delete pod %s/%s at stuck state", pod.Namespace, pod.Name)
+			err = m.deletePod(pod)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (m *StatefulSetControl) deletePod(pod *corev1.Pod) error {
+	return m.Delete(context.TODO(), pod, client.PropagationPolicy(metav1.DeletePropagationBackground))
+}
+
+func (m *StatefulSetControl) getStatefulSetPods(set *appsv1.StatefulSet) ([]*corev1.Pod, error) {
+	selector, err := metav1.LabelSelectorAsSelector(set.Spec.Selector)
+	if err != nil {
+		return nil, err
+	}
+	podList := &corev1.PodList{}
+	err = m.Client.List(context.TODO(), &client.ListOptions{LabelSelector: selector}, podList)
+	if err != nil {
+		return nil, err
+	}
+
+	manager, err := refmanager.New(m.Client, set.Spec.Selector, set, m.scheme)
+	if err != nil {
+		return nil, err
+	}
+	selected := make([]metav1.Object, len(podList.Items))
+	for i, pod := range podList.Items {
+		selected[i] = pod.DeepCopy()
+	}
+	claimed, err := manager.ClaimOwnedObjects(selected)
+	if err != nil {
+		return nil, err
+	}
+
+	claimedPods := make([]*corev1.Pod, len(claimed))
+	for i, pod := range claimed {
+		claimedPods[i] = pod.(*corev1.Pod)
+	}
+	return claimedPods, nil
+}
+
+func calculateStatus(podList []*corev1.Pod, set *appsv1.StatefulSet) (revisionReplicas map[string]*SubsetReplicaStatus) {
+	revisionReplicas = map[string]*SubsetReplicaStatus{}
+	for _, pod := range podList {
+		revision := getRevision(&pod.ObjectMeta)
+		status, exist := revisionReplicas[revision]
+		if !exist {
+			status = &SubsetReplicaStatus{}
+			revisionReplicas[revision] = status
+		}
+		status.Replicas++
+		if podutil.IsPodReady(pod) {
+			status.ReadyReplicas++
+		}
+	}
+	return
+}
+
+func isPodUpgradeComplete(pod *corev1.Pod, revision string) bool {
+	if getRevision(&pod.ObjectMeta) == revision {
+		return podutil.IsPodReadyConditionTrue(pod.Status)
+	}
+	return false
+}
+
+func isPodStuck(pod *corev1.Pod, revision string, partition int32) bool {
+	return !isPodUpgradeComplete(pod, revision) && getOrdinal(pod) >= partition
```

Comment:
The codes here is confusing. I refactored it.

Reply:
It intends to find the stuck pod.
For example. There is a StatefulSet named stateful which pods are READY at beginning.
```
NAME         READY   STATUS    RESTARTS   AGE
stateful-0   1/1     Running   0          93m
stateful-1   1/1     Running   0          93m
stateful-2   1/1     Running   0          93m
```

Then I update the StatefulSet with an incorrect ReadinessProbe. Its pods will look like this:
```
NAME         READY   STATUS    RESTARTS   AGE
stateful-0   1/1     Running   0          93m
stateful-1   1/1     Running   0          93m
stateful-2   0/1     Running   0          4m52s
```

The pod with biggest ordinal will be updated firstly. Because of the wrong Readiness Probe, it will not be READY.

Then I correct the pod template spec by removing the ReadinessProbe. However the pod will not update back. It blocked.

At this situation, these codes try to find this kind of pod and delete it.